configfile: "config.yaml"

from os.path import join, basename, splitext

WD = config['wd']
GENOMES_DIR = join(WD, config['genomes'])
FASTQ_DIR = join(WD, config['fastq'])

WC = glob_wildcards(os.path.join(GENOMES_DIR, "{taxon}.fasta.gz"))

TAXA = WC.taxon
PAIRS = ['R1', 'R2']

rule all:
    input:
        expand("{fastq_dir}/{{taxon}}.{{pair}}.fastq.gz".format(fastq_dir=FASTQ_DIR),
               taxon=TAXA, pair=PAIRS)
    run:
        print("FINISHED WITH NO EXCEPTIONS!")


rule simulate_reads:
    input:
        "{genomes_dir}/{{taxon}}.fasta.gz".format(genomes_dir=GENOMES_DIR)
    output:
        f1 = "{fastq_dir}/{{taxon}}.R1.fastq.gz".format(fastq_dir=FASTQ_DIR),
        f2 = "{fastq_dir}/{{taxon}}.R2.fastq.gz".format(fastq_dir=FASTQ_DIR)
    run:
        output_f1_noext = splitext(output.f1)[0]
        output_f2_noext = splitext(output.f2)[0]
        shell("mkdir -p {fastq_dir}".format(fastq_dir=FASTQ_DIR))
        shell("wgsim -1 100 -2 100 {{input}} {output_f1} {output_f2}".format(output_f1=output_f1_noext, output_f2=output_f2_noext))
        shell("gzip {output_f1} {output_f2}".format(output_f1=output_f1_noext, output_f2=output_f2_noext))
